### PR TITLE
Provide basic macOS support #30

### LINKED ISSUE #30: Provide basic macOS support
Here are the necessary adjustments / considerations to get the project to run on macOS:

- [x] firefoxpwa binary has to be installed to `/usr/local/bin/firefoxpwa`
- [x] firefoxpwa-connector has to be installed to `/usr/local/libexec/firefoxpwa-connector`
- [x] support runtime installation on macOS
  - [x] install dir for macOS is `/usr/local/share/firefoxpwa`
  - [x] runtime executable is `Firefox.app/Contents/MacOS/firefox`
  - [x] runtime config is `Firefox.app/Contents/Resources/application.ini`
  - [x] runtime is extraced from a DMG file
  - [x] macOS target directory for runtime patching is `Firefox.app/Contents/Resources`
- [x] provide manifest for macOS 
- [x] install manifest to `/Library/Application\ Support/Mozilla/NativeMessagingHosts/firefoxpwa.json`

### PR DESCRIPTION
**Pull request by TitanNano**
Closes #30. 

This does not yet update the installation instructions for macOS. 
I'm not sure what the best way for installation on macOS would be. Installing via Homebrew would probably be the most convenient way for many people. I have no idea what submitting this project to Homebrew would involve, though.  Providing a prebuilt ZIP file with all the necessary binaries and assets might also work well enough.

### CODE REVIEW

**Comment by filips123** [comment]
Thanks for the PR!

For installation, I was thinking of using APP/DMG/PKG bundle (I'm not sure what the differences are and which one is appropriate, if any) and/or some of [macOS' package managers](https://apple.stackexchange.com/a/34972). I haven't done much research about this yet, so I'm not sure what doing this would involve, and if it's actually possible.

A few things I'm worried about are permissions (because runtime currently needs to be installed globally, which needs root permissions and I'm not sure how macOS handles that; however, I'm planning to also allow user-specific runtime installation which may solve this issue) and app sandboxing (AFAIK macOS has some app sandbox, and I'm not sure if it conflicts with runtime installation and launching, and using native messaging).

**Comment by TitanNano** [comment]
DMG is just a disc image format and is usually used to distribute GUI applications as they consist of only a single app bundle, which can be dragged and dropped into the global Applications directory of macOS.

APP is a simple directory with a `.app` extensions, which contains all the necessary binaries and static resources that are needed by an Application (icons, translations, images, etc.). It is almost never used by CLI applications, as it's really designed for GUI apps.

PKG can be compared to an MSI on Windows. It's an installation wizard, usually used by more complex applications that require system extensions or something similar and are not just one single app bundle. A PKG could be used for this project to provide a GUI based installation workflow. I think macOS might be required to build a PKG package, though. 

Homebrew is the de facto standard package manager on macOS. It also takes care of the permission issues as `/usr/local` is taken over and managed by Homebrew, so no superuser permissions are required to install or run Homebrew formulas. [Here](https://docs.brew.sh/Formula-Cookbook) is a guide on how to publish something to the repo. 

But as a first step I'd start with letting people build and install from source on macOS and then when everything works out, in a next step, some automatic installation can be figured out.

I will go ahead and update this PR to contain proper `install from source` instructions and fix the CI issues.

**Review by filips123 — CHANGES_REQUESTED** [review]
This PR mostly looks fine, but there are a few changes that should be made. Also, please uncomment [clippy check for macOS](https://github.com/filips123/FirefoxPWA/blob/main/.github/workflows/native.yaml#L29) for GitHub Actions. I will probably merge PR this weekend, because I don't have time until Friday and still need to do a few things.

---

Also, for the record, these features are needed to fully support macOS in the future:

* Packaging (I will probably be able to do this myself)
  * PKG installer (maybe)
  * Homebrew formula (and maybe also packages for other package managers)

* System integration (basically the things that are already in Windows and Linux integrations; I will probably need some help here)
  * When installing PWA site
    * It needs to create an application shortcut/entry with a required icon, name, description, keywords and categories (if macOS doesn't support some of those properties, it's fine, but it needs at least an icon and name) that launches the site.
    * It should also add shortcuts/links to the entry with actions from the manifest. I don't know if this is supported on macOS or how is it called, but on Windows these are jump list tasks and on Linux desktop actions. This functionality is not needed as much as others, so it can be added later.
    * That entry needs to be added to the user-specific app menu/list. I think this could be done by creating a custom `.app` bundle with all those properties that is installed to `~/Applications`.
    * If macOS has some way to list/manage/uninstall apps (such as Add or Remove Programs page in Windows), the site should be registered there, so the user can uninstall it directly from the system.
  * When uninstalling PWA site
    * All system integration entries should be removed.
  * When launching PWA site
    * The window should have a correct site icon. Maybe this can be done in app entry, maybe it also needs some configuration at run-time.
    * The window should not merge/group with other Firefox and PWA windows. On Windows, this is done using Application User Model ID, and on Linux it uses `--class` command argument when launching Firefox to set `WM_CLASS`.
  * *Maybe I also forgot something that is done for Windows and Linux and should also be done here.*

File: native/README.md [hunk 1]
```diff
@@ -115,6 +115,62 @@ sudo chmod 777 /usr/share/firefoxpwa/runtime/
...
+
+# Copy the files to the correct locations
+sudo mkdir -p /usr/local/bin
+sudo install -v target/release/firefoxpwa /usr/local/bin/firefoxpwa
+
+sudo mkdir -p /usr/local/libexec
+sudo install -v target/release/firefoxpwa-connector /usr/local/libexec/firefoxpwa-connector
+
+sudo install -d /Library/Application\ Support/Mozilla/NativeMessagingHosts
+sudo install -v manifests/macos.json /Library/Application\ Support/Mozilla/NativeMessagingHosts/firefoxpwa.json
```
[the remark this answers is not in the document]
Comment (TitanNano) [reply native/README.md:161]: `-D` is not supported on macOS

**Review by filips123 — APPROVED** [review]
I just fixed setting versions, converted the warning to HTML and added some CSS changes.  Now it seems all right. Thanks for the help!

**Pull request merged**